gridded_etl_tools/utils/transform.py

```diff
@@ -0,0 +1,727 @@
+import multiprocessing
+import time
+import json
+import re
+import fsspec
+import pathlib
+import glob
+import os
+import s3fs
+import zarr
+
+from subprocess import Popen
+from typing import Union
+from collections.abc import MutableMapping
+
+import xarray as xr
+
+from kerchunk.hdf import SingleHdf5ToZarr
+from kerchunk.grib2 import scan_grib
+from kerchunk.combine import MultiZarrToZarr
+from tqdm import tqdm
+
+from .convenience import Convenience
+from .metadata import Metadata
+
+TWENTY_MINUTES = 1200
+
+
+class Transform(Metadata, Convenience):
+    """
+    Base class for transforming a collection of downloaded input files first int NetCDF4 Classic format, then
+    (sequentially) kerchunk JSONs, then a MultiZarr Kerchunk JSON,
+    and finally an Xarray Dataset based on that MultiZarr.
+    """
+
+    ##########################
+    # RAW DATA TRANSFORMATIONS
+    ##########################
+
+    # KERCHUNKING
+
+    def create_zarr_json(
+        self, force_overwrite: bool = True, file_filters: list[str] | None = None, outfile_path: str | None = None
+    ):
+        """
+        Convert list of local input files (MultiZarr) to a single JSON representing a "virtual" Zarr
+
+        Read each file in the local input directory and create an in-memory JSON object representing it as a Zarr, then
+        read that collection of JSONs (MultiZarr) into one master JSON formatted as a Zarr and hence openable as a
+        single file
+
+        Note that MultiZarrToZarr will fail if chunk sizes are inconsistent due to inconsistently sized data inputs
+        (e.g. different numbers of steps in input datasets)
+
+        Parameters
+        ----------
+        force_overwrite : bool, optional
+            Switch to use (or not) an existing MultiZarr JSON at `DatasetManager.zarr_json_path()`.
+            Defaults to ovewriting any existing JSON under the assumption new data has been found.
+        file_filters
+            A list of strings used to further filter down input files for kerchunkifying.
+            Useful if you want to kerchunkify only a subset of available files.
+            Defaults to None.
+        outfile_path
+            A custom string path for the final, merged Zarr JSON.
+            Defaults to None, in which case Zarr JSONs are output to self.zarr_json_path.
+        """
+        self.zarr_json_path().parent.mkdir(mode=0o755, exist_ok=True)
+        # Generate a multizarr if it doesn't exist. If one exists, overwrite it unless directed otherwise.
+        if not self.zarr_json_path().exists() or force_overwrite:
+            start_kerchunking = time.time()
+            # Prepapre a list of zarr_jsons and feed that to MultiZarrtoZarr
+            if not hasattr(self, "zarr_jsons"):
```

Comment:
If zarr_jsons is read and written to from outside of this class, we probably should give it a default value at construction time

Reply:
For the time being we've done so for S3 Extractors in the extractor init. @eschechter and I have gone back andforth on this quite a bit though as it's messy. We probably need to figure out a better solution than `zarr_jsons` in the long run

https://github.com/Arbol-Project/gridded-etl-tools/blob/main/gridded_etl_tools/utils/extractor.py#L136

Reply:
This seems related to another comment I wanted to discuss

> I wonder if we should further distinguish between remote and local transformation classes. Remote transformations have made the distinction between transformation and extraction unclear. Remote transformations seem like they need to be arguments to extract, but local transformations can happen independent from extract.

If we for example pass the transform object to the extract object then it seems like we can let transform be in charge of zarr_jsons

Reply:
How do you propose to pass Transform objects to Extract objects? Can you give a pseudocode example?

In general I am strongly skeptical of any attempt to re-combine Transform with either Extract or Publish -- we've spent a lot of effort achieving functional separation as it makes the code a lot easier to write, maintain, and follow. So there'd have to be a good reason born out in actual practice.

Reply:
For sure, what I'm getting at is the same idea. We want transform and extract to be separate, but we see with `zarr_jsons` and I think other properties as well that they are currently tied to each other.

So maybe something like the following could work because it would leave `zarr_jsons` internal to the Transform class.

```
my_etl = CPCPrecipUS()
transformation = Transform(my_etl, Transformation.REMOTE_KERCHUNK)
extract = S3Extractor(my_etl, transformation) # transformation.zarr_jsons available to S3Extractor
```

Reply:
The critical thing here is that `kerchunk` is used for Extract (remotely) and transformation (locally). This means the function and its dependencies (i.e. `zarr_jsons`) pop up in both places and have to be accommodated as such.

So in this case Remote Kerchunk isn't really a tansformation, it's an extraction that makes `transform_data_on_disk` moot since the extracted files are the final files. That seems more logical and easy to follow then instantiating a Remote "transformation" that is not.

Reply:
Looking again at this I'm struggling to see what the marginal gains of this approach are relative to re-writing all our ETLs and transformation stack, which is what it would take. Why is passing transform to extract so advantageous, outside of managing the `zarr_jsons` instance variable more elegantly?

Reply:
To clarify, I'm not proposing we do this in this PR. For this PR, I just think we should consider creating zarr_jsons in the Transform class instead of Extractor, dropping `hasattr`, and then just continuing to let Extract modify zarr_jsons freely for now.

For the approach in general, my reasoning is we're working on refactoring away from DM being a mix-in, so there will need to be a new way of communication between classes. We're taking a step with basic composition, so extract will be able to access zarr_jsons using `self.dm.transform.zarr_jsons`. That's probably an improvement, but I think the goal for any class is for it to work fully independently so that it can be easily reused. If we can further move away from needing `self.dm`, we can work toward having a Transform class that can be applied generally to any data.

About the remote kerchunk, I'm confused because S3Extractor currently calls a request a transformation.

```
def request:
     """
     Transform a remote S3 climate file into a JSON file and add it to the given `DatasetManager` object's internal list of Zarr JSONs.
```

Reply:
Let me rename that docstring, it's confusing, you're right